README.md
# :bulb: dmx-hue

[![NPM version](https://img.shields.io/npm/v/dmx-hue.svg)](https://www.npmjs.com/package/dmx-hue)
[![Build status](https://img.shields.io/travis/sinedied/dmx-hue/master.svg)](https://travis-ci.org/sinedied/dmx-hue)
![Node version](https://img.shields.io/badge/node-%3E%3D6.0.0-brightgreen.svg)
[![XO code style](https://img.shields.io/badge/code_style-XO-5ed9c7.svg)](https://github.com/sindresorhus/xo)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)

![dmx-hue-logo](https://cloud.githubusercontent.com/assets/593151/26761623/710db1ba-4933-11e7-9a08-471e3f9fb9e5.png)

> Art-Net node to control Philips Hue lights with DMX

## Installation

Install [NodeJS](https://nodejs.org), then open a command prompt:

```bash
npm install -g dmx-hue
```

## Usage

```
Usage: dmx-hue [setup] [options]

Create an ArtNet DMX<>Hue bridge.

Options:
  -h, --host       Host address to listen on              [default: '0.0.0.0']
  -a, --address    Set DMX address (range 1-511)          [default: 1]
  -u, --universe   Art-Net universe                       [default: 0]
  -t, --transition Set transition time in ms              [default: 100]
                   Can also be set to 'channel' to enable a dedicated DMX
                   channel on which 1 step equals 100ms.
  -c, --colorloop  Enable colorloop feature
                   When enabled, setting all RGB channels of a light to 1 will
                   enable colorloop mode.
  -n, --no-limit   Disable safety rate limiting
                   Warning: when this option is enabled, make sure to not send
                   more than <number_of_lights>/10 updates per second, or you
                   might overload your Hue bridge.

Note: options overrides settings saved during setup.

Commands:
  setup            Configure hue bridge and DMX options
    -l, --list     List bridges on the network
    -i, --ip       Set bridge IP (use first bridge if not specified)
    --force        Force bridge setup if already configured
```

### Setup

Before being able to control Hue lights on your network, your first have to setup your Hue bridge with the app.
Run `dmx-hue setup`, press the **link button** on your Hue bridge within 30s, then run `dmx-hue setup` again.

After that, follow instructions to configure the default settings.
You can change these settings later at anytime by running `dmx-hue setup` again.

Once your Hue bridge is correctly setup, just run `dmx-hue` to start the Art-Net node, then go crazy with your
favorite DMX controller! :bulb:

If you're looking for a nice cross-platform & open-source DMX software controller, you should take a look at
[QLC+](http://www.qlcplus.org/).

#### Setting lights order

If you need to define your lights order manually for the DMX mapping, edit your configuration file with a text editor
(the file path is displayed at the end of `dmx-hue setup`) and add a new `lightsOrder` entry at the end like this:
```js
{
  ...
  "lightsOrder": [1, 2, 3]  // Hue lights ID
}
```

The Hue lights IDs are displayed when `dmx-hue` is started.

### Hue lights specific features

#### Colorloop

When colorloop option is enabled, your can enable Hue light automatic colorloop mode by setting the R/G/B channels of
a given light to the value `1`.

#### Transition time

Hue light can automatically perform transitions from one state to another in a specified time. This is especially
useful since there is a somewhat low update rate limitation on Hue lights (see next section for more details). If you
want to go creative and be able to adjust transition times dynamically, you can dedicate a DMX channel for it.

## Philips Hue response times vs DMX

With the Philips Hue API it’s only possible to update the state of bulbs 10 times per second, 1 bulb at a time.
Compared to a single DMX universe, which controls all 512 individual parameters up to 44 times per second, there’s
some major differences in how quickly we can update your lights using this software bridge.

This is especially noticeable when using automation sequences, if you try to update Hue lights quicker than 0.1s times
the number of lights in your projects, some updates may be skipped. This is unfortunately a limitation with the Hue
lights API and I cannot do anything about that.

### Safety rate limiting

By default, a safety rate limit is enforced so there is always a 0,1s interval between Hue API calls. You can disable
this limit using the `--no-limit` option, but then you have to make sure to not make more than *number_of_lights / 10*
DMX value changes per second, or your Hue bridge might get overloaded.


CHANGELOG
# 1.1.1

- Fixed DMX adresses starting at 0 instead of 1

# 1.1.0

+ Added option to disable safety rate limiting

# 1.0.1

- Fixed rate limiting issue

# 1.0.0

- Initial version


lib/hue.js
'use strict';

const Color = require('color');
const hue = require('node-hue-api');
const HueApi = hue.api;
const Util = require('./util');

const APP_DESCRIPTION = 'dmx-hue utility';

class Hue {

  getLights() {
    return this.api
      .lights()
      .then(result => result.lights, () => Util.exit('No lights found'));
  }

  createLightState(r, g, b, options) {
    const state = hue.lightState.create().transition(options.transition);
    if (r === g && g === b && b === 0) {
      state.off();
    } else if (options.colorloop && r === g && g === b && b === 1) {
      state
        .on()
        .effect('colorloop')
        .sat(255)
        .bri(255);
    } else {
      const hsv = Color.rgb(r, g, b).hsv().array();
      state.on()
        .effect('none')
        .hsb(hsv[0], hsv[1], hsv[2]);
    }
    return state;
  }

  setLight(id, state) {
    return this.api.setLightState(id, state);
  }

  listBridges(print) {
    return hue
      .nupnpSearch()
      .then(bridges => {
        if (print) {
          bridges.forEach(b => console.log(b.ipaddress));
        }
        return bridges;
      }, () => Util.exit('No bridge found'));
  }

  setupBridge(ip = null, force = false) {
    const bridge = Util.config.get('bridge');
    if (bridge && Util.config.get('user') && !force) {
      console.log(`Bridge configured at ${bridge}`);
      return Promise.resolve();
    }
    return hue
      .nupnpSearch()
      .then(bridges => {
        const bridge = ip ? bridges.find(b => b.ipaddress === ip) : bridges[0];
        if (bridge !== null) {
          Util.config.set('bridge', bridge.ipaddress);
          console.log(`Hue bridge found at ${bridge.ipaddress}`);
          return this.bridge;
        }
        return Promise.reject();
      })
      .catch(() => Util.exit('No bridge found'))
      .then(bridge => new HueApi()
        .registerUser(bridge, APP_DESCRIPTION)
        .then(user => {
          Util.config.set('user', user);
          console.log('Linked bridge successfully');
          return user;
        }, () => Util.exit('Cannot link, press the button on bridge and try again.')));
  }

  get bridge() {
    return Util.config.get('bridge') || Util.exit('Bridge not configured, run "hue-dmx setup"');
  }

  get user() {
    return Util.config.get('user') || Util.exit('Bridge not linked, run "hue-dmx setup"');
  }

  get api() {
    if (!this._api) {
      this._api = new HueApi(this.bridge, this.user);
    }
    return this._api;
  }

}

module.exports = Hue;


lib/artnet.js
'use strict';

const dgram = require('dgram');
const Util = require('./util');
const Network = require('./network');

const NODE_NAME = 'HueDMX';
const PORT = 6454;
const HEADER = 'Art-Net\0';
const OPCODE_OUTPUT = 0x5000;     // ArtDMX data packet
const OPCODE_POLL = 0x2000;       // ArtPoll packet
const OPCODE_POLLREPLY = 0x2100;  // ArtPollReply packet
const MIN_PROTOCOL_VERSION = 14;

class ArtNet {

  static listen(address, dmxHandler) {
    // Set up the socket
    const socket = dgram.createSocket('udp4', (msg, peer) => {
      const header = msg.toString('utf8', 0, 8);
      const opcode = msg.readUInt16LE(8);
      const version = msg.readUInt16BE(10);

      if (header !== HEADER || version < MIN_PROTOCOL_VERSION) {
        return;
      }

      if (opcode === OPCODE_OUTPUT) {
        // Deserialize DMX data
        const sequence = msg.readUInt8(12);
        const physical = msg.readUInt8(13);
        const universe = msg.readUInt16LE(14);
        const length = msg.readUInt16BE(16);
        const dmx = [];

        for (let i = 0; i < length; ++i) {
          dmx.push(msg.readUInt8(18 + i));
        }

        dmxHandler({sequence, physical, universe, length, dmx}, peer);
      } else if (opcode === OPCODE_POLL) {
        // Send ArtPollReply
        const buffer = Buffer.alloc(238);
        const itf = Network.interface();
        const ip = itf.address.split('.').map(s => parseInt(s, 10));
        const mac = itf.mac.split(':').map(s => parseInt(s, 16));
        buffer.write(HEADER, 0, 8);
        buffer.writeUInt16LE(OPCODE_POLLREPLY, 8);
        for (let i = 0; i < 4; i++) {
          buffer[10 + i] = ip[i];
        }
        buffer.writeUInt16LE(PORT, 14);
        buffer[23] = 0xA0;                // Status 1
        buffer.write('tm', 24, 2);        // ESTA code
        buffer.write(NODE_NAME, 26, 18);  // Short name
        buffer.write(NODE_NAME, 44, 64);  // Long name
        buffer[173] = 1;                  // Number of ports
        buffer[174] = 0x0f;               // Port 1 can output DMX
        buffer[183] = 0x0f;               // Port 1 good output
        buffer[190] = 0;                  // Port 1 address
        for (let i = 0; i < 6; i++) {
          buffer[201 + i] = mac[i];
        }
        buffer[212] = 0xC;                // Status 2

        const to = address ? address.split('.')[0] + '.255.255.255' : '255.255.255.255';
        socket.setBroadcast(true);
        socket.send(buffer, 0, buffer.length, PORT, to);
      }
    });

    return new Promise(resolve => {
      socket.on('error', err => Util.exit('Error: ' + (err.message ? err.message : err)));
      socket.bind(PORT, address, () => resolve(socket));
    });
  }
}

module.exports = ArtNet;


index.js
'use strict';

const minimist = require('minimist');
const inquirer = require('inquirer');
const chalk = require('chalk');
const Util = require('./lib/util');
const Hue = require('./lib/hue');
const ArtNet = require('./lib/artnet');
const pkg = require('./package.json');

const help =
`${chalk.bold('Usage:')} dmx-hue [setup] [options]

Create an ArtNet DMX<>Hue bridge.

${chalk.bold('Options:')}
  -h, --host       Host address to listen on              [default: '0.0.0.0']
  -a, --address    Set DMX address (range 1-511)          [default: 1]
  -u, --universe   Art-Net universe                       [default: 0]
  -t, --transition Set transition time in ms              [default: 100]
                   Can also be set to 'channel' to enable a dedicated DMX
                   channel on which 1 step equals 100ms.
  -c, --colorloop  Enable colorloop feature
                   When enabled, setting all RGB channels of a light to 1 will
                   enable colorloop mode.
  -n, --no-limit   Disable safety rate limiting
                   Warning: when this option is enabled, make sure to not send
                   more than <number_of_lights>/10 updates per second, or you
                   might overload your Hue bridge.

Note: options overrides settings saved during setup.

${chalk.bold('Commands:')}
  setup            Configure hue bridge and DMX options
    -l, --list     List bridges on the network
    -i, --ip       Set bridge IP (use first bridge if not specified)
    --force        Force bridge setup if already configured
`;

class DmxHue {

  constructor(args) {
    this._args = minimist(args, {
      boolean: ['list', 'force', 'help', 'version', 'colorloop', 'no-limit'],
      string: ['ip', 'host', 'transition'],
      number: ['address', 'universe'],
      alias: {
        l: 'list',
        i: 'ip',
        h: 'host',
        a: 'address',
        t: 'transition',
        c: 'colorloop',
        n: 'no-limit',
        u: 'universe'
      }
    });
    this._hue = new Hue();
    this._lastUpdate = 0;
    this._delayedUpdate = null;
  }

  start(options) {
    if (options.address <= 0 || options.address > 511) {
      Util.exit('Invalid DMX address');
    }

    options = Object.assign({}, options);
    this._hue
      .getLights()
      .then(lights => {
        const ordered = options.order.reduce((r, id) => {
          const light = lights.find(light => light.id === id.toString());
          if (light && !options.disabled[id]) {
            r.push(light);
          }
          return r;
        }, []);
        const remaining = lights.filter(light => !options.disabled[light.id] && !ordered.find(o => light.id === o.id));
        options.lights = ordered.concat(remaining);
        options.transitionChannel = options.transition === 'channel';
        options.colors = {};
        const dmxChannelCount = (3 * options.lights.length) + (options.transitionChannel ? 1 : 0);
        const extraDmxAddress = options.address + dmxChannelCount - 512;

        if (extraDmxAddress >= 0) {
          console.warn(chalk.yellow('Warning: not enough DMX channels, some lights will be unavailable'));
          const lightsToRemove = Math.ceil(extraDmxAddress / 3.0);
          options.lights = options.lights.slice(0, -lightsToRemove);
        }

        return ArtNet.listen(options.host, data => {
          if (data.universe === options.universe) {
            this._updateLights(data.dmx, options);
          }
        });
      })
      .then(() => {
        let currentAddress = options.address;
        if (options.noLimit) {
          console.warn(chalk.yellow('Warning, safety rate limiting is disabled!\n'));
        }
        console.log(chalk.bold(`DMX addresses on universe ${options.universe}:`));
        if (options.transitionChannel) {
          console.log(` ${currentAddress++}: transition time`);
        }
        options.lights.forEach(light => {
          console.log(` ${chalk.cyan(`${currentAddress}:`)} ${light.name} ${chalk.grey(`(Hue ID: ${light.id})`)}`);
          currentAddress += 3;
        });
        console.log('\nArtNet node started (CTRL+C to quit)');
      });
  }

  setup(ip, force) {
    this._hue.setupBridge(ip, force).then(() => this._setupOptions());
  }

  _updateLights(dmxData, options) {
    if (this._delayedUpdate) {
      clearTimeout(this._delayedUpdate);
      this._delayedUpdate = null;
    }

    let address = options.address - 1;

    if (options.transitionChannel) {
      options.transition = dmxData[address] * 100;
      address++;
    }

    const dmx = dmxData.slice(address, address + (3 * options.lights.length));
    let j = 0;

    for (let i = 0; i < options.lights.length; i++) {
      const lightId = options.lights[i].id;
      const color = dmx.slice(j, j + 3);
      const previous = options.colors[lightId];
      j += 3;

      // Update light only if color changed
      if (!previous || color[0] !== previous[0] || color[1] !== previous[1] || color[2] !== previous[2]) {
        // Rate limit Hue API to 0,1s between calls
        const now = new Date().getTime();
        if (options.noLimit || now - this._lastUpdate >= 100) {
          const state = this._hue.createLightState(color[0], color[1], color[2], options);
          this._lastUpdate = now;
          this._hue.setLight(lightId, state);
          options.colors[lightId] = color;
        } else if (!this._delayedUpdate) {
          // Make sure to apply update later if changes could not be applied
          this._delayedUpdate = setTimeout(() => this._updateLights(dmxData, options), 100);
        }
      }
    }
  }

  _setupOptions() {
    const disabled = Util.config.get('disabledLights') || {};
    const transition = Util.config.get('transition') || 0;
    this._hue
      .getLights()
      .then(lights => {
        return inquirer
          .prompt([
            {
              type: 'input',
              name: 'dmxAddress',
              message: 'Set DMX address (range 1-511)',
              default: Util.config.get('dmxAddress') || 1,
              validate: input => {
                const value = parseInt(input, 10);
                return value > 0 && value <= 511;
              }
            },
            {
              type: 'input',
              name: 'universe',
              message: 'Set Art-Net universe',
              default: Util.config.get('universe') || 0,
              validate: input => parseInt(input, 10) > 0
            },
            {
              type: 'confirm',
              name: 'colorloop',
              message: 'Enable colorloop feature',
              default: Util.config.get('colorloop') || false
            },
            {
              type: 'confirm',
              name: 'transitionChannel',
              message: 'Use DMX channel for transition time',
              default: transition === 'channel'
            },
            {
              type: 'input',
              name: 'transition',
              message: 'Set transition time in ms',
              default: transition === 'channel' ? 100 : transition,
              when: answers => !answers.transitionChannel,
              validate: input => parseInt(input, 10) > 0
            },
            {
              type: 'checkbox',
              name: 'lights',
              message: 'Choose lights to use',
              choices: lights.map(light => ({
                name: light.name,
                value: light.id,
                checked: !disabled[light.id]
              }))
            }
          ])
          .then(answers => {
            Util.config.set('dmxAddress', parseInt(answers.dmxAddress, 10));
            Util.config.set('universe', parseInt(answers.universe, 10));
            Util.config.set('colorloop', answers.colorloop);
            Util.config.set('transition', answers.transitionChannel ? 'channel' : parseInt(answers.transition, 10));
            Util.config.set('disabledLights', lights
              .filter(light => !answers.lights.includes(light.id))
              .reduce((l, light) => {
                l[light.id] = true;
                return l;
              }, {}));
            console.log(`Configuration saved at ${chalk.green(Util.config.path)}`);
          });
      });
  }

  run() {
    if (this._args.help) {
      Util.exit(help, 0);
    } else if (this._args.version) {
      Util.exit(pkg.version, 0);
    } else if (this._args._[0] === 'setup') {
      return this._args.list ? this._hue.listBridges() : this.setup(this._args.ip, this._args.force);
    }

    if (this._args.transition !== 'channel') {
      const value = parseInt(this._args.transition, 10);
      this._args.transition = value >= 0 ? value : 0;
    }

    return this.start({
      host: this._args.host,
      address: this._args.address || Util.config.get('dmxAddress') || 1,
      colorloop: this._args.colorloop || Util.config.get('colorloop') || false,
      transition: this._args.transition || Util.config.get('transition') || 100,
      noLimit: this._args['no-limit'] || Util.config.get('noLimit') || false,
      universe: this._args.universe || Util.config.get('universe') || 0,
      disabled: Util.config.get('disabledLights') || {},
      order: Util.config.get('lightsOrder') || []
    });
  }

}

module.exports = DmxHue;
